lib/pbench/test/functional/server/test_put.py

```diff
@@ -44,44 +44,68 @@ def test_upload_all(self, server_client: PbenchServerClient, login_user):
+        return not_indexed, indexed
+
     def test_index_all(self, server_client: PbenchServerClient, login_user):
         """Wait for datasets to reach the "Indexed" state, and ensure that the
         state and metadata look good
         """
+        print(" ... reporting behaviors ...")
 
         # Test get_list pagination: to avoid forcing a list, we'll count the
         # iterations separately. (Note that this is really an implicit test
         # of the paginated datasets/list API and the get_list generator, which
         # one could argue belong in a separate test case; I'll likely refactor
         # this later when I add GET tests.)
         count = 0
-        datasets = server_client.get_list(limit=5)
+        not_indexed = server_client.get_list(limit=5, metadata=["server.tarball-path"])
 
         # For each dataset we find, poll the state until it reaches Indexed
         # state, or until we time out.
         start = time.time()
         timeout = start + (60.0 * 10.0)
 
-        for dataset in datasets:
-            count += 1
-            while True:
-                metadata = server_client.get_metadata(
-                    dataset.resource_id, ["dataset.state", "server.status"]
+        while True:
+            now = time.time()
+            print(f"[{now - start:0.2f}] Checking ...")
+            not_indexed, indexed = TestPut.check_indexed(server_client, not_indexed)
+            for dataset in indexed:
+                count += 1
+                print(
+                    "    Indexed ",
+                    dataset.metadata["server.tarball-path"],
                 )
-                state = metadata["dataset.state"]
-                status = metadata["server.status"]
-                stats = set(status.keys()) if status else set()
-                if state == "Indexed" and {"unpack", "index"} <= stats:
-                    # Don't wait for backup, and don't fail if we haven't as
-                    # it's completely independent from unpack/index; but if we
-                    # have backed up, check that the status was successful.
-                    if "backup" in stats:
-                        assert status["backup"] == "ok"
-                    assert status["unpack"] == "ok"
-                    assert status["index"] == "ok"
-                    break
-                assert (
-                    time.time() < timeout
-                ), f"Exceeded timeout: {dataset.name} state {state}, status {status}"
+            if not not_indexed:
+                break
+            if time.time() < timeout:
                 time.sleep(30.0)  # sleep for half a minute
-        assert count == len(self.tarballs), "Didn't find all expected datasets"
+            else:
+                print("Exceeded timeout (10 minutes), still not indexed:")
+                for dataset in not_indexed:
+                    print(dataset.metadata["server.tarball-path"])
+        assert count == len(
```

Comment:
It seems like there is a `break` or `assert` statement missing here:  if we time out, how do we exit the (infinite) loop at line 91?

If indeed that needs fixing, then I have a few more suggestions:
```
        while not_indexed:
            now = time.time()
            if now < timeout:
                time.sleep(30.0)  # sleep for half a minute
            else:
                assert False, (
                    f"Timed out after {now - start} seconds; unindexed datasets: "
                    + ", ".join(d.metadata["server.tarball-path"] for d in not_indexed)
                )
            print(f"[{now - start:0.2f}] Checking ...")
            not_indexed, indexed = TestPut.check_indexed(server_client, not_indexed)
            for dataset in indexed:
                count += 1
                print("    Indexed ", dataset.metadata["server.tarball-path"])
```
(That is:  replace the loop exit condition and frontload the wait, use a single value of "now", and put the timeout message into the assertion.)

Reply:
I'll post a follow-up PR shortly.

Reply:
See PR #3142.